planner/core/indexmerge_path.go

```diff
@@ -510,131 +510,252 @@ func (ds *DataSource) generateAndPruneIndexMergePath(indexMergeConds []expressio
-				indexMergeIsIntersection = true
-				jsonPath = sf.GetArgs()[0]
-				var ok bool
-				//virCol.RetType
-				vals, ok = jsonArrayExpr2Exprs(ds.ctx, sf.GetArgs()[1], virCol.GetType())
-				if !ok {
-					continue
-				}
-			case ast.JSONOverlaps: // (json_overlaps(a->'$.zip', '[1, 2, 3]')
-				var jsonPathIdx int
-				if sf.GetArgs()[0].Equal(ds.ctx, targetJSONPath) {
-					jsonPathIdx = 0 // (json_overlaps(a->'$.zip', '[1, 2, 3]')
-				} else if sf.GetArgs()[1].Equal(ds.ctx, targetJSONPath) {
-					jsonPathIdx = 1 // (json_overlaps('[1, 2, 3]', a->'$.zip')
-				} else {
-					continue
-				}
-				jsonPath = sf.GetArgs()[jsonPathIdx]
-				var ok bool
-				vals, ok = jsonArrayExpr2Exprs(ds.ctx, sf.GetArgs()[1-jsonPathIdx], virCol.GetType())
-				if !ok {
-					continue
-				}
-			default:
-				continue
-			}
+		indexMergePath := &util.AccessPath{PartialIndexPaths: partialPaths}
+		indexMergePath.IndexMergeIsIntersection = isIntersection
+		indexMergePath.TableFilters = remainingFilters
 
-			// Step 2.2. Check some limitations.
-			if jsonPath == nil || len(vals) == 0 {
-				continue
-			}
-			if !jsonPath.Equal(ds.ctx, targetJSONPath) {
-				continue // not on the same JSON col
-			}
+		// TODO: use a naive estimation strategy here now for simplicity, make it more accurate.
+		minEstRows, maxEstRows := math.MaxFloat64, -1.0
+		for _, p := range indexMergePath.PartialIndexPaths {
+			minEstRows = math.Min(minEstRows, p.CountAfterAccess)
+			maxEstRows = math.Max(maxEstRows, p.CountAfterAccess)
+		}
+		if indexMergePath.IndexMergeIsIntersection {
+			indexMergePath.CountAfterAccess = minEstRows
+		} else {
+			indexMergePath.CountAfterAccess = maxEstRows
+		}
 
-			// Step 2.3. Generate a IndexMerge Path of this filter on the current MVIndex.
-			var partialPaths []*util.AccessPath
-			for _, v := range vals {
-				partialPath := &util.AccessPath{Index: mvIndex}
-				partialPath.Ranges = ranger.FullRange()
-				// TODO: get the actual column length of this virtual column
-				partialPath.IdxCols, partialPath.IdxColLens = []*expression.Column{virCol}, []int{types.UnspecifiedLength}
-				partialPath.FullIdxCols, partialPath.FullIdxColLens = []*expression.Column{virCol}, []int{types.UnspecifiedLength}
-
-				// calculate the path range with the condition `a->'$.zip' = 1`.
-				eq, err := expression.NewFunction(ds.ctx, ast.EQ, types.NewFieldType(mysql.TypeTiny), virCol, v)
-				if err != nil {
-					return nil, err
-				}
-				if err = ds.detachCondAndBuildRangeForPath(partialPath, []expression.Expression{eq}); err != nil {
-					return nil, err
-				}
+		mvIndexPaths = append(mvIndexPaths, indexMergePath)
+	}
+	return
+}
+
+func (ds *DataSource) buildPartialPaths4MVIndex(accessFilters []expression.Expression,
+	idxCols []*expression.Column, mvIndex *model.IndexInfo) ([]*util.AccessPath, bool, error) {
+	var virColID = -1
+	for i := range idxCols {
+		if idxCols[i].VirtualExpr != nil {
```

Comment:
what if the composite index contains two virtual columns?
`create table t(a int, b int as (a + 1) virtual, c json, index idx(b, (cast(c as signed array))));`

then the index contains two virtual columns, one is `b` and another is a hidden column

Reply:
Does the user have use cases like that? If not, I'll forbid this case. 
I want to make the scope as small as possible in this delivery to decrease risks.